controllers/utils.go

```diff
@@ -65,61 +65,79 @@ func mergeMap(target map[string]string, desired map[string]string) bool {
 }
 
-func extendArg(container *corev1.Container, defaultArgs map[string]string) (args map[string]string, err error) {
-	flags := internal.FlagSet{}
-	if err = flags.Parse(container.Args); err != nil {
-		return
-	}
+func extendArgs(args, externalArgs []string) []string {
+	containerArgFlags := internal.FlagSet{}
+	_ = containerArgFlags.Parse(args)
 
-	// flag in the args doesn't contain prefix '-' or '--'
-	args = flags.Flags()
-	for flag, value := range defaultArgs {
-		// we need to cut the prefix '-' or '--' before comparing with existingVars
-		flag = strings.TrimLeft(flag, "-")
-		if _, ok := args[flag]; !ok {
-			args[flag] = value
+	existKeys := make(map[string]struct{}, len(containerArgFlags.Flags()))
+	for flag := range containerArgFlags.Flags() {
+		key := strings.TrimLeft(flag, "-")
+		if flag == "-p" {
+			key = "port"
 		}
+		existKeys[key] = struct{}{}
 	}
 
-	container.Args = make([]string, 0, len(args)*2)
-	// sort the arg list
-	flags.Visit(func(flag, value string) {
-		container.Args = append(container.Args, "--"+flag)
-		if value != "" {
-			container.Args = append(container.Args, value)
+	externalArgFlags := internal.FlagSet{}
+	_ = externalArgFlags.Parse(externalArgs)
+
+	for flag, value := range externalArgFlags.Flags() {
+		key := strings.TrimLeft(flag, "-")
+		if _, ok := existKeys[key]; !ok {
+			containerArgFlags.Flags()[flag] = value
 		}
+	}
+
+	mergedArgs := []string{}
+	containerArgFlags.Visit(func(flag, value string) {
+		mergedArgs = append(mergedArgs, flag)
+		mergedArgs = append(mergedArgs, value)
 	})
-	return
+	return mergedArgs
 }
 
-func extendPorts(args map[string]string, userDefinedPorts, defaultPorts []corev1.ContainerPort) []corev1.ContainerPort {
-	// copy default ports and cover the containerPort with user-defined port arg
-	required := coverPorts(args, defaultPorts)
-	// merge user-defined ports to required
-	return mergePorts(required, userDefinedPorts)
+func extendPorts(ports, externalPorts []corev1.ContainerPort) []corev1.ContainerPort {
+	for i := range externalPorts {
+		found := false
+		for j := range ports {
+			if (&ports[j]).Name == (&externalPorts[i]).Name {
+				found = true
+				break
+			}
+		}
+		if !found {
+			if (&externalPorts[i]).Name == "" {
+				(&externalPorts[i]).Name = fmt.Sprintf("unset-%d", (&externalPorts[i]).ContainerPort)
+			}
+			ports = append(ports, externalPorts[i])
+		}
+	}
+	return ports
 }
 
 // coverPorts use the port in user-defined args to cover the default port
-func coverPorts(args map[string]string, required []corev1.ContainerPort) []corev1.ContainerPort {
-	ports := make([]corev1.ContainerPort, len(required))
-	copy(ports, required)
+func coverPortsFromArgs(args []string, ports []corev1.ContainerPort) []corev1.ContainerPort {
+	flags := internal.FlagSet{}
+	_ = flags.Parse(args)
 
-	for i := range required {
-		if port, ok := args[(&required[i]).Name]; ok {
+	for i := range ports {
+		name := (&ports[i]).Name
+		if port, ok := flags.Flags()["--"+name]; ok {
```

Comment:
Using a temporary variable to refer to the flags instead of calling func to get the return value every time in foreach, this is a good programming practice.